File: sse-pkg/util/bintail.cpp
/*******************************************************************************

 File:    bintail.cpp
 Project: OpenSonATA
 Authors: The OpenSonATA code is the result of many programmers
          over many years

 Copyright 2011 The SETI Institute

 OpenSonATA is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 OpenSonATA is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with OpenSonATA.  If not, see<http://www.gnu.org/licenses/>.
 
 Implementers of this code are requested to include the caption
 "Licensed through SETI" with a link to setiQuest.org.
 
 For alternate licensing arrangements, please contact
 The SETI Institute at www.seti.org or setiquest.org. 

*******************************************************************************/

/* 
 * Similar to 'tail -f' but for binary files.
 * Takes filename as first arg, sends output to stdout.
 * At EOF, program sleeps until more input is available.
 * If input file is truncated, the program exits.
 * Written in C to get nonblocking reads to work.
 */

#include <cstdio>
#include <cstdlib>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#define BUFF_SIZE 32768

int getFileSize(int filedes)
{
    int filesize = 0;

    struct stat buf;
    if (fstat(filedes, &buf) == 0)
    {
        filesize = buf.st_size;
    }
    else 
    {
       fprintf(stderr, "error finding filesize for filedes %d\n",
	       filedes);
    }

    return filesize;
}


int main(int argc, char *argv[])
{
   char buf[BUFF_SIZE];
   long pos;
   int fd;
   bool debug(false);

   if (argc < 2)
   {
      fprintf(stderr, "usage: prog <filename>\n");
      exit(EXIT_FAILURE);
   }

   char *fname = argv[1];

   if (-1 == (fd = open(fname, O_RDONLY | O_NONBLOCK, 0)))
   {
      fprintf(stderr,"Cannot open file: %s\n", fname);
      exit(EXIT_FAILURE);
   }

   if (-1 == lseek(fd, 0L, SEEK_SET))  // seek to beginning
   {
      fprintf(stderr,"Cannot lseek file: %s\n", fname);
      exit(EXIT_FAILURE);
   }

   if ((pos = lseek(fd, 0L, SEEK_CUR)) == -1)
   {
      fprintf(stderr,"Cannot seek file: %s\n", fname);
      exit(EXIT_FAILURE);
   }


   bool alreadyReadSomeData = false;

   for (;;)
   {
      //fgets(buf, sizeof(buf), fp);
      int readCount = read(fd, buf, sizeof(buf));

      if (debug)
      {
	 fprintf(stderr, "readcount is %d\n", readCount);
      }

      if (readCount == -1)
      {
	 // no data available, sleep

	 if (debug)
	 {
	    fprintf(stderr, "sleeping...\n");
	 }

	 sleep(1); /* give up the cpu to other processes */

	 continue;
      }

      if (readCount == 0)
      {
	 // eof
	 if (debug)
	 {
	    fprintf(stderr, "at eof\n");
	 }

	 if (alreadyReadSomeData & getFileSize(fd) == 0)
	 {
	    if (debug)
	    {
	       fprintf(stderr, "file was truncated\n");
	    }
	    break;
	 }

	 // At EOF, remember this file position, and
	 // sleep, waiting for more data

	 lseek(fd, pos, SEEK_SET);

	 if (debug)
	 {
	    fprintf(stderr, "sleeping...\n");
	 }
	 sleep(1); /* give up the cpu to other processes */

	 continue;

      }

      alreadyReadSomeData = true;

      pos = lseek(fd, 0L, SEEK_CUR);

      if (debug)
      {
	 fprintf(stderr, "read %d\n", readCount);
      }

      write(STDOUT_FILENO, buf, readCount);


   }

}

File: sse-pkg/jseeker/BinTail.java
/*******************************************************************************

 File:    BinTail.java
 Project: OpenSonATA
 Authors: The OpenSonATA code is the result of many programmers
          over many years

 Copyright 2011 The SETI Institute

 OpenSonATA is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 OpenSonATA is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with OpenSonATA.  If not, see<http://www.gnu.org/licenses/>.
 
 Implementers of this code are requested to include the caption
 "Licensed through SETI" with a link to setiQuest.org.
 
 For alternate licensing arrangements, please contact
 The SETI Institute at www.seti.org or setiquest.org. 

*******************************************************************************/

package opensonata.dataDisplays;

// tail a binary file, passing data to stdout.
// If file is at EOF, sleep until more input is available.
// If file is truncated, then exit.

import java.io.*;
import java.text.*;

public class BinTail {

    // Read file containing binary data.
    // Effectively does a 'tail -f' on the file, sleeping periodically
    // until new data appears.
    // If an EOFException is received, it means the file has been truncated
    // and it's time to exit.

   DataInputStream in;
   byte[] byteArray;

   BinTail() {
      int MAX_READBYTES = 512;

      byteArray = new byte[MAX_READBYTES];
   }

   private void openFile(String filename)
   {
      try {
	 in = new DataInputStream(
	    new BufferedInputStream(
	       new FileInputStream(filename)));
      }
      catch (IOException e)
      {
	 System.err.println(e);
	 System.exit(1);
      }
   }
   
   private void sleep()
   {
      try {
	 //System.err.println("sleeping");
	 int latency = 200;  //mS
	 
	 Thread.sleep(latency);
      }
      catch (InterruptedException e) {
	 // Interrupt may be thrown manually by stop()
      }
   }


   public void run() {

      boolean continueRunning = true;
      
      while (continueRunning) {
	 
	 try {

	    // Keep reading data, and passing it to stdout
		    
	    int x = 0;
	    while (in != null && in.available() != 0) {

	       // read it
	       int nread = in.read(byteArray);

	       System.err.println("read " + nread + " bytes");
	       if (nread == -1)
	       {
		  // file was truncated
		  System.err.println("file was truncated, exiting");
		  System.exit(0);
	       }

	       // write it
	       int offset = 0;
	       System.out.write(byteArray, offset, nread);

	    }
	 }
	 catch (EOFException e)
	 {
	    // Input file was truncated.  
	    // Reopen file to start over.

	    System.err.println(e);

	    System.exit(0);

	 }
	 catch (IOException e)
	 {
	    System.err.println(e);

	 }

	 //System.err.println("read " + count + " bytes");

	 // wait for more data to appear
	 //sleep();
      }
   }

   public static void main(String[] args) {

      if (args.length < 1)
      {
	 System.err.println("usage: bintail <filename>");
	 System.exit(1);
      }

      System.err.println("started reading...");
      
      BinTail binTail = new BinTail();

      binTail.openFile(args[0]);
      binTail.run();
      
   }
}